File: Data Structures & Algorithms/house-robber/submission-3.cpp
#include<bits/stdc++.h>
class Solution {
public:
    int n;
    int t[101];
    int solve(vector<int>& nums, int index) {
        if (index >= n) return 0;
        if (t[index]!= -1) return t[index];
        int pick = nums[index] + solve(nums,index+2);
        int skip = solve(nums, index+1);
        return t[index] = max(pick, skip);
    }
    int rob(vector<int>& nums) {
        n = nums.size();
        memset(t, -1, sizeof(t));
        int result = solve(nums, 0);
        return result;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    solve(nums, index) {
        let val = 1;
        for (let i = 0; i < nums.length; i++) {
            if (index == i) continue;
            val *= nums[i];
        }

        return val;
    }
    productExceptSelf(nums) {
        const result = [];
        let value;
        for (let i = 0; i < nums.length; i++) {
            value = this.solve(nums, i);
            result.push(value);
        }

        return result;
    }
}
